src/circuit_proof/mod.rs

```diff
@@ -0,0 +1,781 @@
+#![allow(non_snake_case)]
+
+use rand::{CryptoRng, Rng};
+use std::iter;
+
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use curve25519_dalek::traits::{IsIdentity, MultiscalarMul, VartimeMultiscalarMul};
+
+use generators::Generators;
+use inner_product_proof::{inner_product, InnerProductProof};
+use proof_transcript::ProofTranscript;
+use util;
+
+#[derive(Clone, Debug)]
+pub struct Circuit {
+    n: usize,
+    m: usize,
+    q: usize,
+    c: Vec<Scalar>,        // vector of constants, length q
+    W_L: Vec<Vec<Scalar>>, // q vectors, of length n each
+    W_R: Vec<Vec<Scalar>>,
+    W_O: Vec<Vec<Scalar>>,
+    W_V: Vec<Vec<Scalar>>, // q vectors, of length m each
+}
+
+impl Circuit {
+    pub fn check_parameters(&self) -> Result<(), &'static str> {
+        if self.W_L.len() != self.q
+            || self.W_R.len() != self.q
+            || self.W_O.len() != self.q
+            || self.W_V.len() != self.q
+        {
+            return Err("Circuit matrix size doesn't match specified parameters.");
+        }
+        if self.c.len() != self.q {
+            return Err("Circuit constants vector size doesn't match specified parameters.");
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct CircuitProof {
+    pub A_I: RistrettoPoint,
+    pub A_O: RistrettoPoint,
+    pub S: RistrettoPoint,
+    pub T_1: RistrettoPoint,
+    pub T_3: RistrettoPoint,
+    pub T_4: RistrettoPoint,
+    pub T_5: RistrettoPoint,
+    pub T_6: RistrettoPoint,
+    pub t_x: Scalar,
+    pub t_x_blinding: Scalar,
+    pub e_blinding: Scalar,
+    pub ipp_proof: InnerProductProof,
+}
+
+impl CircuitProof {
+    pub fn prove<R: Rng + CryptoRng>(
+        gen: &Generators,
```

Comment:
The existing `Generators` struct is designed to hold generators for an `m`-aggregated `n`-bit rangeproof, and requires that `m` and `n` both be powers of two.

But here, the `Generators` struct isn't being used that way, it's being used to hold generators for a single circuit with `n` multiplication gates, and `n` isn't required to be a power of two.

Reply:
Hmm, you're right that the `Generators` struct doesn't check that `n, m` are powers of two. I think that's a bug.

My concern is that the `Generators` struct is designed to be used for `m`-aggregated `n`-bit rangeproofs, and here we're not doing that.  So, I think that we should either use a different type for circuit generators (which requires thinking about how we want to relate the circuit generators to the circuit description), or change the `Generators` struct so that it's not just for rangeproofs (which requires checking that this doesn't have any other effects on the rest of the code).